### DISCUSSION TITLE
Trouble shooting on low latency mode not being used

### DISCUSSION THREAD

**Discussion by 8128** [google/oboe#1955] [General]
In the Q&A part `I requested a stream with PerformanceMode::LowLatency, but didn't get it. Why not?`
It mentioned that there are several possibilities - 
1. No data callback
2. Not a good sample rate
3. Used AudioFormat::Float below android 9.0
4. Device does not support low latency stream
5. Requested a channel count which is not supported natively by the audio device
6. Maximum number of LowLatency streams has been reached
7. You are on Android 7.0 or below

I'm debugging on the reason of me not getting low latency mode, and I've set callback(met reqirement 1), set default sample rate [here](https://github.com/google/oboe/blob/master/docs/GettingStarted.md#obtaining-optimal-latency)(met requirement 2), I'm also using android 10(met 3, 7), and using mono channel(should be good for 5)

It can be still possible that 4, 6 are causing the issue, the question is how can I know if my device support low latency or not? Also is there a way to query if maximum number of LowLatency streams has been reached or not? 

Thanks in advance.

**Comment by robertwu1** [comment]
Which device are you on?

Can you try reproing the issue with OboeTester?

**Comment by philburk** [comment]
> set default sample rate [here](https://github.com/google/oboe/blob/master/docs/GettingStarted.md#obtaining-optimal-latency)(met requirement 2)

On some devices, input and output have different optimal rates! Ie output may run optimally at 48000 and input at 44100!

Try setting everything UNSPECIFIED except for PerformanceMode and callback. The AAudio should find the optimal settings.

What device are you using? Some devices do not support LowLatency!

How are you detecting not having LowLatency? Are you calling getPerformanceMode()? Or just hearing high latency?

**Comment by 8128** [comment]
Thank you @philburk, I've changed the rates to unspecified but I'm still getting high latency. Also, I cannot download OboeTester @robertwu1 mentioned since I'm on a Android TV device - I'm actually using NVIDIA SHIELD TV Pro and I'm trying to get the input from the USB port it has. I checked `getPerformanceMode()` and I'm sure it is not low latency, in the test the latency is about 350ms.

I used Chromecast 4k with Google TV as well, with a usb hub, and the input there can get in with performance mode - although the latency is still over 150ms, it is way better than this NVIDIA SHIELD TV Pro, I guess it's just a device issue.

**Reply by robertwu1** [reply]
If you enable developer mode and USB debugging, you can install OboeTester manually from your laptop via an USB cable. A local version of OboeTester can be found [here](https://github.com/google/oboe/tree/main/apps/OboeTester).

Many TVs do not support low latency so I suspect this is WAI.
